### ISSUE TITLE
Build CNN using NumPy and Convolve over MNIST - Visualize Feature Map

### ISSUE THREAD

**Issue by iphton** [numpy/numpy#12387]
I like to build a basic CNN forward block using NumPy and use it on MNIST dataset. To build a simple CNN block ( convolution, relu, max_pool ), I define some following function. This function basically adopted from online though and with bit modification by me.

**Define Padding**

```python
def zero_pad(X, pad):
	'''
	X: python numpy array shape (m, n_H, n_W, n_C)
	pad: integer, amount of padding around each image

	X_pad : padding image (m, n_H + 2*pad , n_W + 2*pad, n_C)

	'''
	return np.pad(X, ((0,0) , (pad, pad) , (pad, pad) , (0,0)) , 'constant' , constant_values = 0)

```

Then single convolve. 

```python
def conv_single_step(a_slice_prev, W, b):
	# element wise multiplication
	return np.sum(np.multiply(a_slice_prev, W) + b)
```

Then full convolution forward pass program that will invoke above `conv_single_step` function for each step.

```python
def conv_forward(A_prev, W, b, hparameters):
	# retrive dimensions from A_prev's shape
	(m, n_H_prev, n_W_prev, n_C_prev) = A_prev.shape

	# retrieve dimensions from W's shape
	(f, f, n_C_prev, n_C) = W.shape

	# retrieve information from hparameters 
	stride = hparameters['stride']
	pad = hparameters['pad']

	# compute the dimensions of the conv output
	n_H = int((n_H_prev - f + 2*pad)/stride) + 1
	n_W = int((n_W_prev - f + 2*pad)/stride) + 1

	# initialize the output volume 
	Z = np.zeros((m, n_H, n_W, n_C))

	A_prev_pad = zero_pad(A_prev, pad)

	for i in range(m):
		a_prev_pad = A_prev_pad[i]
		for h in range(n_H):
			for w in range(n_W):
				for c in range(n_C):
					vert_start = h*stride
					vert_end = vert_start + f 
					horiz_start = w*stride
					horiz_end = horiz_start + f

					# use the corner to define the 3D slice 
					a_slice_prev = a_prev_pad[vert_start:vert_end, horiz_start:horiz_end, :]

					# convolve
					Z[i, h , w, c] = conv_single_step(a_slice_prev, W[...,c], b[...,c])

	assert(Z.shape == (m, n_H, n_W, n_C))

	cache = (A_prev, W, b, hparameters)

	return Z, cache	

```

Then a relu activation function.

```python
def relu(A_prev):  
	
	# retrive dimensions from A_prev's shape
	(m, n_H_prev, n_W_prev, n_C_prev) = A_prev.shape
	# initialize the output volume 
	A = np.zeros((m, n_H_prev, n_W_prev, n_C_prev))

	for i in range(m):  
		for h in numpy.arange(0, n_H_prev):  
			for w in numpy.arange(0, n_W_prev):  
				for c in numpy.arange(0, n_C_prev):
					A[i, h, w, c] =  np.maximum(A[i, h, w, c], 0)  
	return A 

```

After this We define pooling layer.

```python
def pool_forward(A_prev, hparameters, mode = 'max'):
	(m, n_H_prev, n_W_prev, n_C_prev) = A_prev.shape

	f = hparameters['f']
	stride = hparameters['stride']

	n_H = int(1 + (n_H_prev - f) / stride)
	n_W = int(1 + (n_W_prev - f) / stride)
	n_C = n_C_prev

	# Initialize the output matrix
	A = np.zeros((m, n_H, n_W, n_C))

	for i in range(m):
		for h in range(n_H):
			for w in range(n_W):
				for c in range(n_C):

					vert_start = h*stride
					vert_end = vert_start + f
					horiz_start = w*stride
					horiz_end = horiz_start + f

					# use the corner to define the current slice
					a_prev_slice = A_prev[i, vert_start:vert_end, horiz_start:horiz_end, c]

					if mode == 'max':
						A[i, h, w, c] = np.max(a_prev_slice)
					elif mode == 'average':
						A[i, h, w, c] = np.mean(a_prev_slice)

	cache = (A_prev, hparameters)
	assert(A.shape == (m, n_H, n_W, n_C))

	return (A, cache)

```

Now, we've create very basic convolutional forward building block. The MNIST data set is split into two part: **train** and **test**.

```python

train = pd.read_csv("train.csv")
test = pd.read_csv("test.csv")

Y_train = train["label"]

# Normalize the data
X_train = X_train / 255.0
test = test / 255.0

# Reshape image in 3 dimensions 
X_train = X_train.values.reshape(-1,28,28,1)

# Randomly split the data sets
from sklearn.model_selection import train_test_split
X_train, X_val, Y_train, Y_val = train_test_split(X_train, Y_train, test_size = 0.1,
                                                  random_state=random_seed)


W = np.random.randn(2, 2, 1, 8)
b = np.random.randn(1, 1, 1, 8)
hparameters = {"pad" : 2,
               "stride": 1}

Z, cache_conv = conv_forward(X_train, W, b, hparameters )
A_prev = relu(Z)
hparameters = {"stride" : 1, "f": 4}

A, cache = pool_forward(A_prev, hparameters)

print('Shape of Z : ', Z.shape)  # (37800, 31, 31, 8)
print('Shape of A_prev : ', A_prev.shape) # (37800, 31, 31, 8)
print('Shape of A : ', A.shape) # (37800, 28, 28, 8)

```

Now, I want to visualize each feature map, but can't get it.

```python

import matplotlib.pyplot as plt
a = A[:,:,:,0]
a.shape
plt.imshow(a[0][:,:])
```

Any suggestion will be highly appreciated. Thanks.

**Comment by charris** [comment]
Questions like this are best asked on the NumPy mailing list: https://mail.python.org/mailman/listinfo/numpy-discussion. The github issues are reserved for bug reports and suggestions.
